lib/sdk.js

```diff
@@ -0,0 +1,414 @@
+        .then(() => extractSdk(checksumVerify, path.basename(url), download(url)));
+    });
+};
+
+exports.installRustlib = () => {
+  var url = null;
+  var checksumVerify = null;
+  var rustv = null;
+  var pkgname = null;
+
+  return exports.rustVersion()
+    .then(_rustv => {
+      rustv = _rustv;
+      pkgname = 'MIPS libstd v' + rustv;
+      url = rustlibUrl.replace('VERSION', rustv);
+
+      return downloadString(url + '.sha256');
+    })
+    .catch(() => {
+      throw new Error(`Could not find a MIPS libstd matching your current Rust version (${rustv}). Only stable Rust versions >= 1.11.0 are supported.`);
+    })
+    .then(checksum => {
+      checksumVerify = checksum;
+      return exports.checkRustlib(rustv, checksumVerify);
+    })
+    .then(check => {
+      if (check.exists && check.checked) {
+        log.info(`Latest ${pkgname} already installed.`);
+        return;
+      } else if (!check.exists) {
+        log.info(`Installing ${pkgname}...`);
+      } else {
+        log.info(`Updating ${pkgname}...`);
+      }
+
+      return fs.mkdirpAsync(paths.rustlib)
+        .then(() => extractRustlib(checksumVerify, path.basename(url), download(url), rustv));
+    });
+};
+
+function extract(checksumVerify, filename, sdkStream, root, strip, name, decompress) {
+  log.info(`Downloading ${name}...`);
+
+  return tmpdir()
+    .then(destdir => {
+      // Exract tarball to destination.
+      var extract = tar.extract(destdir.path, {
+        strip: strip,
+        ignore: function(name) {
+          // Ignore self-directory.
+          return path.normalize(name + '/') === path.normalize(destdir.path + '/');
+        }
+      });
+
+      function tryCleanup() {
+        try {
+          destdir.cleanup();
+        } catch (e) {}
+      }
+
+      return new Promise((resolve, reject) => {
+        var checksum = '';
+        sdkStream
+          .pipe(sha256stream())
+          .on('sha256', function(sha256) {
+            checksum = sha256file(sha256, filename);
+          })
+          .pipe(decompress)
+          // tar-stream has a recursion issue when input chunks are too big.
+          // by splitting up the chunks, we never get too deeply nested in the
+          // stack.
+          .pipe(blocks({
+            size: 64 * 1024,
+            zeroPadding: false
+          }))
+          .pipe(extract)
+          .on('finish', () => {
+            // Check sum.
+            if (checksum !== checksumVerify) {
+              return reject(new Error(`Checksum for downloaded ${name} does not match!`));
+            }
+
+            // Write out CHECKSUM file.
+            fs.writeFileSync(path.join(destdir.path, 'CHECKSUM'), checksum);
+
+            // Remove the old SDK directory.
+            fs.removeAsync(root)
+              .then(() => {
+                // Move temporary directory to target destination.
+                return fs.moveAsync(destdir.path, root);
+              })
+              .finally(() => {
```

Comment:
t2-cli should use https://www.npmjs.com/package/promise.prototype.finally (which is based on actual proposal)